### PR TITLE
Update the Freshmaker Docker Image to Fedora 34

### PR DESCRIPTION
**Pull request by ronnyhlim**
Updating the Docker image to use Fedora 34. Once merged in Dev, will run the Ansible tower job and see if there are any issues.

### CODE REVIEW

File: Dockerfile
Comment (gnaponie) [inline Dockerfile:1]: I guess you could remove this :)

**Review by gnaponie — COMMENTED** [review]
It looks good. Of course test it in dev as you mentioned before releasing it to prod.

**Comment by qixiang** [comment]
> Updating the Docker image to use Fedora 34. Once merged in Dev, will run the Ansible tower job and see if there are any issues.
> 
> https://docs.google.com/...

This is a public repo, so please not share internal access only links in this repo.

File: Dockerfile
Comment (qixiang) [inline Dockerfile:2]: In additional to updating this, I believe there are more work to be done to run freshmaker with fedora 34. At least, I believe there are some changes required to work with the `python3-flask-migrate` in fedora 34, I can't remember the exact reason, you can try running db migrations to check this, probably there are more changes required to support the newer version of flask components.

**Review by qixiang — COMMENTED** [review]
And you need to install python3-rhmsg for fedora 34 from somewhere, currently it's installed from my copr repo which only has the fedora 32 build.

File: Dockerfile
Comment (ronnyhlim) [reply Dockerfile:2]: @qixiang, I'm not familiar with what a copr repo is; Can I just follow these instructions (https://docs.pagure.org/copr.copr/user_documentation.html) to install python3-rhmsg on my copr?

Also, regarding the changes needed for "python3-flask-migrate", are the instructions in _docs/db.rst_ things I can test for after the PR gets merged? or is this something I must test beforehand? 

Lastly, I don't have permission to merge the repo and I don't see anywhere on Github if this was approved or not. How can I tell if this PR was marked approved? (On Pagure it was just a +1 and on Gitlab it would say approved)

Comment (qixiang) [reply Dockerfile:2]: > Can I just follow these instructions (https://docs.pagure.org/copr.copr/user_documentation.html) to install python3-rhmsg on my copr?

Yes, it should be more easier to do that just with web UI (https://copr.fedorainfracloud.org/coprs/), you need to create a project and build `python3-rhmsg` in your project with the Fedora 34 chroot.

> Also, regarding the changes needed for "python3-flask-migrate", are the instructions in docs/db.rst things I can test for after the PR gets merged? or is this something I must test beforehand?

That's a good start point to test the migrations with the instructions in `docs/db.rst`, but we can't merge this before identifying the broken functionalities and fixing them.

> Lastly, I don't have permission to merge the repo and I don't see anywhere on Github if this was approved or not.
 
Probably you need to be in `redhat-exd-rebuilds` organization, @gnaponie should have the permission to add you to the org or grant you the merge permission to this project.

> How can I tell if this PR was marked approved? (On Pagure it was just a +1 and on Gitlab it would say approved)

Github PR also has "Approved", we add +1 in comments, for freshmaker, usually we need one +1 for minor changes, and better to ask 2 or more +1 for large or complicated changes.

Comment (ronnyhlim) [reply Dockerfile:2]: ahh sweet okay, I'm starting to understand the bits a little better. So I forked your existing "python-rhmsg" from your COPR page and changed the chroot to Fedora 34 x86_64 (https://copr.fedorainfracloud.org/coprs/rlim/python3-rhmsg/); I'm trying to submit the build but I'm not sure what the "URLs to file" should be.

Also, once the build is submitted and the Dockerfile is changed to use the new COPR build with f34, I can then follow the steps in "docs/db.rst" to find any other changes needed correct? ie changes in "python3-flask-migrate"

Comment (gnaponie) [reply Dockerfile:2]: @ronnyhlim I've grant you access. Let me know if it works. @qixiang I also made you an owner, you should be able to change permission too now.

Comment (qixiang) [reply Dockerfile:2]: > ahh sweet okay, I'm starting to understand the bits a little better. So I forked your existing "python-rhmsg" from your COPR page and changed the chroot to Fedora 34 x86_64 (https://copr.fedorainfracloud.org/coprs/rlim/python3-rhmsg/); I'm trying to submit the build but I'm not sure what the "URLs to file" should be.

I've no idea about "URLs to file", it has been a while since last time I used copr, a more easier way to start a build could be:
```
copr-cli build rlim/python3-rhmsg /path/to/local/source/rpm/file
```
You can download the source rpm from
https://download.copr.fedorainfracloud.org/results/mikeb/python-rhmsg/fedora-33-x86_64/01690645-python-rhmsg/python-rhmsg-0.10.1-2.fc33.src.rpm
or 
https://download.copr.fedorainfracloud.org/results/qwan/python-rhmsg/fedora-32-x86_64/01680268-python-rhmsg/python-rhmsg-0.10.1-2.fc32.src.rpm

the dist name (fc32, fc33) in source rpm file doesn't matter, make sure you have the fedora 34 x86_64 build root enabled in your copr project. Sorry, I'm suggesting you to go back to play with the cli :-/  

> Also, once the build is submitted and the Dockerfile is changed to use the new COPR build with f34, I can then follow the steps in "docs/db.rst" to find any other changes needed correct? ie changes in "python3-flask-migrate"

That's one of things you can check to verify freshmaker code is compatible with the newer version of packages, but not the only thing to check, however I can't tell which APIs or functionalities will be broken with the new version of packages, I'd suggest you at least run freshmaker locally with some events and verify the handlers are working and call the web APIs to verify they're still working.

Comment (ronnyhlim) [reply Dockerfile:2]: > @ronnyhlim I've grant you access. Let me know if it works. @qixiang I also made you an owner, you should be able to change permission too now.

@gnaponie Confirming that I've received access. Thank you!

Comment (ronnyhlim) [reply Dockerfile:2]: @qixiang  Thanks for the links and instructions! I was able to successfully use the CLI to finish the build. I have 2 questions

1) general question: I remember it was mentioned that setting up the dev environment for Freshmaker wasn't so easy with a Dockerfile. I personally found Dockerfile to be easier to use but wasn't sure if there was something I'm missing as to why Vagrant was preferred. Is there an issue with testing if the API works by launching a Docker container and running the API? Or should be I be using Vagrant?

2) Within the "db.rst" file I'm seeing that I need to install tons of modules [example: "ModuleNotFoundError: No module named 'flask_migrate'] when I run the first step to set up the database "_./manager created_". Am I supposed to run these commands within the docker container (sudo docker exec –it fmaker_cont /bin/bash)? Or do I just run them locally after installing the many modules that are required?

Comment (qixiang) [reply Dockerfile:2]: > @qixiang Thanks for the links and instructions! I was able to successfully use the CLI to finish the build. I have 2 questions
> 
> 1. general question: I remember it was mentioned that setting up the dev environment for Freshmaker wasn't so easy with a Dockerfile. I personally found Dockerfile to be easier to use but wasn't sure if there was something I'm missing as to why Vagrant was preferred. Is there an issue with testing if the API works by launching a Docker container and running the API? Or should be I be using Vagrant?

If you only need to run the freshmaker consumer (which runs the fedmsg-hub), a single container could be sufficient and easy, however sometimes you may want to run the frontend and the fedmsg-relay as well for testing the web APIs, then a single container might not be a good choice, you will need some some other solutions like docker compose to run multiple containers, but that introduced extra effort to configure the `fedmsg-relay` and database to work with the multiple containers setup. Both freshmaker frontend and backend will need to access the database and when some APIs are called, the frontend will publish messages to the `manual.rebuild` topic, `fedmsg-relay` should forward these messages, so the backend consumer can receive these messages. Though you may find running all these components in a single container is also possible, but that is not the recommended way of using containers and not so flexible.

This doesn't mean you cannot use container for setting up your local dev environment, if that works for you, just feel free to use that, I recommend vagrant just because I personally think it's easier and not error prone.

> 2. Within the "db.rst" file I'm seeing that I need to install tons of modules [example: "ModuleNotFoundError: No module named 'flask_migrate'] when I run the first step to set up the database "_./manager created_". Am I supposed to run these commands within the docker container (sudo docker exec –it fmaker_cont /bin/bash)? Or do I just run them locally after installing the many modules that are required?

In case your container image has been built with all the dependency RPMs installed which was done by the
```
RUN \
    dnf -y install $(cat /tmp/yum-packages.txt) python3-rhmsg findutils && \
    dnf clean all
```
this may be something changed in new version of flask and flask-migrate. If you're running the migrations in container, then yes, you need to verify the relevant rpms are installed correctly in the container. In this case, using vagrant is more easier to uninstall/install, upgrade/downgrade these RPMs for troubleshooting, since container doesn't persist your changes if you don't want to rebuild your image for every changes.

Comment (ronnyhlim) [reply Dockerfile:2]: @qixiang hmm I see, but if I'm not using the Dockerfile/COPR build, but using Vagrant instead (with "vagrant box add fedora/34-cloud-base"), then what is the use of the Dockerfile/COPR build? I don't see the vagrant file point to the Dockerfile anywhere. Apologies if this is a silly question, this is my first time using Vagrant and am slightly confused.

In other words, how would I test the change in the Dockerfile as requested in the task if I'm using vagrant to test the change instead? Would I test using Vagrant (with f34) and assume the issues there would be the same ones from the Dockerfile?

Comment (qixiang) [reply Dockerfile:2]: > @qixiang hmm I see, but if I'm not using the Dockerfile/COPR build, but using Vagrant instead (with "vagrant box add fedora/34-cloud-base"), then what is the use of the Dockerfile/COPR build?

If you're installing the RPMs in vagrant VM with the same instructions in Dockerfile, that won't be a problem, the purpose is verifying Freshmaker can work with new version of relevant RPMs in Fedora 34, it doesn't matter you're using a Fedora 34 vagrant VM or a container. If this makes any confusion, and you're more comfortable with using container for the testing, feel free to test with container.

> I don't see the vagrant file point to the Dockerfile anywhere.

Yes, these are independent choice of setting up the running environment. You can play with VMs, vagrant VMs, containers or even your host OS system, the only thing you need to care about is the environment should be same with our prod environment (which is running in container, currently Fedora 32, the dependencies are rpms installed via dnf), so no matter what's the solution you're using, make sure you're running with the same os release, rpm versions.

> Apologies if this is a silly question, this is my first time using Vagrant and am slightly confused.

No worries :)

> In other words, how would I test the change in the Dockerfile as requested in the task if I'm using vagrant to test the change instead? Would I test using Vagrant (with f34) and assume the issues there would be the same ones from the Dockerfile?

Yes, if you identify any issues with f34 vagrant VM that breaks freshmaker, that would also happen with f34 container, and you need some fixes for that.

**Review by qixiang — APPROVED** [review]
Did some testing locally with dry run builds, though without UMB + rhmsg, so far it looks good.

**Pull request merged**
